### PR TITLE
[GSoC] Added MNIST Sample Exporter

### PR DESCRIPTION
**Pull request by captain-pool**
CC: @vbardiovskyg

### CODE REVIEW

File: examples/mnist/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Sample MNIST Graph to export as Tensorflow Hub Module
+------------------------------------------------------
+### To Export
+```bash
+$ python3 export.py
+```
+### to run The Tests
+- `$ bazel test E1_TFHub_Sample_Deploy:export_test`
```
Comment (vbardiovskyg) [inline examples/mnist/README.md]: Please update this.

File: examples/mnist/export.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
+import sys
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import tensorflow_hub as hub
+import argparse
+""" Create a Sample TF-Hub Module using SavedModel v2.0
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: I would rewrite this as:

"""Exporter for TF-Hub Modules in SavedModel v2.0 format.

The module has as a single signature, accepting a batch of images with shape [None, 28, 28, 1] and returning a prediction vector.

In this example, we are loading the MNIST Dataset from TFDS and training a simple digit classifier.
"""

File: examples/mnist/export.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+The module has as a single signature which loads MNIST Dataset from TFDS and train a simple Neural Network for classifying the digits. The model is built and trained using Tewnsorlfow
+"""
+FLAGS = None
+
+
+class MNIST(tf.keras.models.Model):
+  """Model class for MNIST Classifier
+  """
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: """ should go on the end of the previous line

File: examples/mnist/export.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import tensorflow_hub as hub
+import argparse
+""" Create a Sample TF-Hub Module using SavedModel v2.0
+The module has as a single signature which loads MNIST Dataset from TFDS and train a simple Neural Network for classifying the digits. The model is built and trained using Tewnsorlfow
+"""
+FLAGS = None
+
+
+class MNIST(tf.keras.models.Model):
+  """Model class for MNIST Classifier
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Let's call it:

"""Module representing a MNIST classifier."""

File: examples/mnist/export.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  """
+
+  def __init__(self, output_activation="softmax"):
+    """
+    Args:
+      output_activation (str): activation for last layer
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Period at the end.

File: examples/mnist/export.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    super(MNIST, self).__init__()
+    self.layer_1 = tf.keras.layers.Dense(64)
+    self.layer_2 = tf.keras.layers.Dense(10, activation=output_activation)
+
+  @tf.function(
+      input_signature=[
+          tf.TensorSpec(
+              shape=[
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: cannot the shape definition go on one line?

shape=[None, 28, 28, 1]

File: examples/mnist/export.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    flatten = tf.keras.layers.Flatten()(casted)
+    normalize = tf.keras.layers.Lambda(
+        lambda x: x / tf.reduce_max(tf.gather(x, 0)))(flatten)
+    x = self.layer_1(normalize)
+    output = self.layer_2(x)
+    return output
+
+
+def train_step(model, loss_fn, optimizer_fn, metric, image, label):
+  """ Trains the model
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: """Perform one training step for the model.

  Args:
    ...

File: examples/mnist/export.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    Args:
+      model: Keras Model to train
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Use periods please.

File: examples/mnist/export.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      loss_fn: Loss Function to use
+      optimizer_fn: Optimizer function to use
+      metric: keras.metric to use
+      image: shape[batch_size, width, height, num_channels] Tensor of Training Images
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: better would be:

image: Tensor of training images of shape [batch_size, 28, 28, 1].

Same with the other arg definitions.

File: examples/mnist/export.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      label: shape[batch_size,] Image classes as returned by TFDS
+  """
+  with tf.GradientTape() as tape:
+    preds = model(image)
+    label_onehot = tf.one_hot(label, 10)
+    loss_ = loss_fn(label_onehot, preds)
+  grads = tape.gradient(loss_, model.trainable_variables)
+  optimizer_fn.apply_gradients(zip(grads, model.trainable_variables))
+  metric(loss_)
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Now it could be my misunderstanding of the API, but isn't this call returning a tensor that you have to later use?

File: examples/mnist/export.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        buffer_size=1000,
+        batch_size=32,
+        epoch=10,
+        export_path="/tmp/tfhub_modules/mnist/digits/1"):
+  """
+    Trains and export the Model as SavedModel 2.0
+    Args:
+      data_dir (str): Directory where to store datasets from TFDS (With proper authentication, Cloud Bucket Supported)
+      buffer_size (int): Size of Buffer to use while shuffling
+      batch_size (int): size of each training batch
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Please either use Capital first letter or not, but consistently across the whole file. Additionally, please use periods at the ends of sentences everywhere.

File: examples/mnist/export.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+      export_path (str): path to export the trained model
+  """
+  model = MNIST()
+  kwargs = {}
+  if data_dir:
+    kwargs["data_dir"] = data_dir
+  train = tfds.load("mnist", split="train", **kwargs)
+  optimizer_fn = tf.optimizers.Adam(learning_rate=1e-3)
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: shouldn't this be tf.keras.optimizers.Adam? In that case, you might not need the learning_rate as default is the same.

File: examples/mnist/export.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  loss_fn = tf.keras.losses.CategoricalCrossentropy(from_logits=True)
+  metric = tf.keras.metrics.Mean()
+  model.compile(optimizer_fn, loss=loss_fn)
+  train = train.shuffle(
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: please apply this shuffle and batch where you do tfds.load, all in one statement.

File: examples/mnist/export_test.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import os
+from absl.testing import absltest
+import export
+from export import train_and_export
+
+TMPDIR = tempfile.mkdtemp()
+model = None
+TEST_DATA_DIR = "examples/mnist/test_data"
+
+def load_model(path):
```
Comment (vbardiovskyg) [inline examples/mnist/export_test.py]: Please remove this and just use hub.load(path)

File: examples/mnist/export_test.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+import tensorflow as tf
+import tensorflow_hub as hub
+import unittest
+import tempfile
+import shutil
+import os
+from absl.testing import absltest
+import export
+from export import train_and_export
```
Comment (vbardiovskyg) [inline examples/mnist/export_test.py]: Please don't import functions, but modules instead:
from examples.mnist import export

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#22-imports

File: examples/mnist/export.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    Args:
+      data_dir (str): Directory where to store datasets from TFDS (With proper authentication, Cloud Bucket Supported)
+      buffer_size (int): Size of Buffer to use while shuffling
+      batch_size (int): size of each training batch
+      export_path (str): path to export the trained model
+  """
+  model = MNIST()
+  kwargs = {}
+  if data_dir:
+    kwargs["data_dir"] = data_dir
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: Why? Just pass in None if not provided, it will result in the same behavior.

File: examples/mnist/export.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  metric(loss_)
+
+
+def train_and_export(
+        data_dir=None,
+        buffer_size=1000,
+        batch_size=32,
+        epoch=10,
+        export_path="/tmp/tfhub_modules/mnist/digits/1"):
+  """
```
Comment (vbardiovskyg) [inline examples/mnist/export.py]: """Trains and export the Model as SavedModel 2.0.

  Args:
    ...

File: examples/mnist/export_test.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+class TFHubMNISTTest(tf.test.TestCase):
+  def setUp(self):
+    file_ = tf.gather(tf.io.matching_files("%s/*.jpg" % TEST_DATA_DIR), 0)
+    self.test_image = tf.expand_dims(
+        tf.image.decode_jpeg(
+            tf.io.read_file(file_)), 0)
+    self.test_label = int(str(file_.numpy()).split("/")[-1].split(".jpg")[0])
+
+  def test_model_exporting(self):
+    mocked_dict = export.sys.__dict__.copy()
```
Comment (vbardiovskyg) [inline examples/mnist/export_test.py]: What's going on here? Not sure what this test is doing.

File: examples/mnist/export_test.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+      mocked_dict["stdout"] = devnull
+      absltest.mock.patch.dict(
+          export.sys.__dict__,
+          mocked_dict,
+          clear=True).start()
+      train_and_export(epoch=1, export_path="%s/model/1" % TMPDIR)
+      self.assertTrue(os.listdir(TMPDIR))
+
+  def test_empty_input(self):
+    if not os.path.exists("%s/model/1" % TMPDIR):
```
Comment (vbardiovskyg) [inline examples/mnist/export_test.py]: Instead of TMPDIR, please use self.get_temp_dir()

Also remove this condition and export always into newly create self.get_temp_dir() with train_and_export.

File: examples/mnist/export_test.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+      self.test_model_exporting()
+    model = load_model("%s/model/1" % TMPDIR)
+    prediction = model.call(self.test_image)
+    self.assertEqual(
+        tf.argmax(
+            tf.squeeze(prediction)).numpy(),
+        self.test_label)
+
+  @classmethod
+  def tearDownClass(cls):
```
Comment (vbardiovskyg) [inline examples/mnist/export_test.py]: Not needed in tf.test framework.

File: .gitignore [hunk 21]
```diff
@@ -0,0 +1,15 @@
+# Compiled Bytes
+__pycache__/
```
Comment (vbardiovskyg) [inline .gitignore]: Remove this file from the commit.

File: examples/mnist/export.py (hunk 12 above)
Comment (captain-pool) [reply examples/mnist/export.py]: tf.keras.optimizers.Adam and tf.optimizers.Adam are actually the same. yep, I forgot about that learning_rate parameter.

File: examples/mnist/README.md (hunk 1 above)
Comment (captain-pool) [reply examples/mnist/README.md]: Can you rephrase? didn't really get what you said.

File: examples/mnist/export.py (hunk 13 above)
Comment (captain-pool) [reply examples/mnist/export.py]: Done :)

File: examples/mnist/export.py (hunk 16 above)
Comment (captain-pool) [reply examples/mnist/export.py]: Oops. Done :)

File: examples/mnist/export.py (hunk 17 above)
Comment (captain-pool) [reply examples/mnist/export.py]: Done :)

File: examples/mnist/README.md (hunk 1 above)
Comment (vbardiovskyg) [reply examples/mnist/README.md]: E1_TFHub_Sample_Deploy has no meaning here.

File: .gitignore (hunk 21 above)
Comment (captain-pool) [reply .gitignore]: Without the gitignore, it is adding the bazel symlinks on git add -A
Else, I need to add the files one by one

File: examples/mnist/export_test.py (hunk 18 above)
Comment (captain-pool) [reply examples/mnist/export_test.py]: This test is training the model for one iteration and exporting the graph. The reason for this test is to check if the trained model is being exported properly or not.

File: examples/mnist/export.py (hunk 10 above)
Comment (captain-pool) [reply examples/mnist/export.py]: There's an option either to use the Tensor being generated, or I can just call, `result()` which will give me the same thing.
`tensor = metric(loss)`
and
`metric(loss)
tensor = metric.result()`
returns same values :)

File: examples/mnist/export_test.py (hunk 18 above)
Comment (vbardiovskyg) [reply examples/mnist/export_test.py]: I meant the mocking of sys.__dict__. I have a strong feeling that this is not needed.

File: examples/mnist/export_test.py (hunk 19 above)
Comment (captain-pool) [reply examples/mnist/export_test.py]: Done :)

File: examples/mnist/export_test.py (hunk 20 above)
Comment (captain-pool) [reply examples/mnist/export_test.py]: Done :)

File: examples/mnist/export_test.py (hunk 15 above)
Comment (captain-pool) [reply examples/mnist/export_test.py]: Thanks for sharing the link :)

File: examples/mnist/export_test.py (hunk 14 above)
Comment (captain-pool) [reply examples/mnist/export_test.py]: I'm using this avoid reloading the model everytime. This would reduce the test time. at first I did `hub.load()` in `def setUp(...)` but turns out, at the beginning of each and every step, all the instance variables are cleared and `setUp` is called. So, it is reloading the same model every single time. That's reducing the speed of the test.

**Comment by captain-pool** [comment]
CC: @vbardiovskyg

File: .gitignore (hunk 21 above)
Comment (vbardiovskyg) [reply .gitignore]: You can keep the file locally, just remove it from the commit.

File: examples/mnist_export_v2/export_test.py [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  def setUp(self):
+    file_ = tf.gather(tf.io.matching_files("%s/*.jpg" % TEST_DATA_DIR), 0)
+    self.test_image = tf.expand_dims(
+        tf.image.decode_jpeg(
+            tf.io.read_file(file_)), 0)
+    self.test_label = int(str(file_.numpy()).split("/")[-1].split(".jpg")[0])
+
+  def test_model_exporting(self):
+    mocked_dict = export.sys.__dict__.copy()
+    with open(os.devnull, "w") as devnull:
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export_test.py]: Can you test the exporting behaviour without mocking?

File: examples/mnist_export_v2/export_test.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+import unittest
+import tempfile
+import os
+from absl.testing import absltest
+import export
+
+model = None
+TEST_DATA_DIR = "examples/mnist_export_v2/test_data"
+
+def load_model(path):
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export_test.py]: I understand the speed reduction, but loading should be extremely fast in this and doesn't justify adding a confusing global variable.

Also, what if we want to export and load a different model every time?

File: examples/mnist_export_v2/export_test.py (hunk 22 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: Yep, I can actually! But, I mocked it to prevent printing anything. Sachin said, while unittesting it is better to not print anything. Is it okay to print out the training steps while testing?

Comment (vbardiovskyg) [reply examples/mnist_export_v2/export_test.py]: I see. I think printing is not ideal, but still better than unreadable code. You can make it train for minimal number of steps in the test scenario, so there won't be much printing.

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: For this test we are looking for one trained model. The idea is, it will train the model in `test_model_exporting`. And this trained model will be loaded for checking if this can actually be used for inferencing. If I remove this, it will train the model on the same dataset, thrice for three different steps, just to obtain  th trained graph. Once for exporting, once for testing on empty_input, and finally once for inferencing.

Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: I'm not really sure, if removing this would help.

File: examples/mnist_export_v2/export_test.py (hunk 22 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: Okay! On it! This should just print one line. As it is clearing out the previous line using `\r` :)

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (vbardiovskyg) [reply examples/mnist_export_v2/export_test.py]: Yes, I see. It looks like a not very good design to do the following things:
1. call one test from another,
2. make one test call produce artifacts used by a different test.

It's fine to create it multiple times. You can create it the smallest possible and train for 1 step.

Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: Do you think training for one step would produce enough updates to run the inference test?

Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: Hey @vbardiovskyg , I just found out, since I've been using `tmpfile.mkdtemp()`, there was this unavoidable need to use `hub.load()`. since after you asked me to use `self.get_temp_dir()`, this problem of file getting deleted after each step, is solved! So, removing that `load_model()` function would work just fine! Thanks :)

File: .gitignore (hunk 21 above)
Comment (captain-pool) [reply .gitignore]: Removed :)

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (vbardiovskyg) [reply examples/mnist_export_v2/export_test.py]: Honestly, I think the most reasonable way is to not test the trained model for whether it can predict an image with reasonable accuracy.

Instead I would just test for formal correctness, meaning that the code is syntactically right, etc. This way you don't need to include any test data / images and we will save a lot of resources. At the risk of messing up this script in the future without noticing, but that risk is extremely low.

File: examples/mnist/README.md (hunk 1 above)
Comment (captain-pool) [reply examples/mnist/README.md]: Yep. Fixed :)

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: So, what do you suggest? should I remove that? I noticed something similar in text_embeddings
https://github.com/tensorflow/hub/blob/caa9e03f3b483fa611efd49a81b5c512ef62a6dc/examples/text_embeddings/export_test.py#L66-L84
So, I thought, why not test if the model is even producing something which actually makes sense. If you suggest, I'll remove this test :), but I think, being a small enough model, and having a test data of just a few kilobytes, it shouldn't consume too much of resource. and also bazel test currently takes 57 secs to run.

**Comment by captain-pool** [comment]
CC: @vbardiovskyg

File: examples/mnist_export_v2/export_test.py [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+        tf.image.decode_jpeg(
+            tf.io.read_file(file_)), 0)
+    self.test_label = int(str(file_.numpy()).split("/")[-1].split(".jpg")[0])
+
+  def test_model_exporting(self):
+    export.train_and_export(epoch=1, batch_size=64, export_path="%s/model/1" % self.get_temp_dir())
+    self.assertTrue(os.listdir(self.get_temp_dir()))
+
+  def test_empty_input(self):
+    if not os.path.exists("%s/model/1" % self.get_temp_dir()):
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export_test.py]: Please don't call another test from a test and don't rely on anything produced by a different test.

Same below.

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (vbardiovskyg) [reply examples/mnist_export_v2/export_test.py]: The difference is that exporter in text_embeddings is not training anything, just building a simple graph.

I think it is totally fine to test training on a couple of images and then run inference. I would even recommend that. But I don't think it is fine to train a model fully to achieve a predictable performance.

File: examples/mnist_export_v2/export_test.py (hunk 24 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: `test_empty_input()` is loading the graph exported during the test and checking if it is running on empty input. So, as a matter of fact it does depend on the test model exported during the training. Can you suggest me some ways to get around this dependency? One way might be to look if the graph exists and if it doesn't, I can call that train_and_export function again from the test. However `test_model_exporting()` exactly does this thing. so to make things modular I did this. Even i think the tests shouldn't be interdependent, and I'm open to suggestions to get around this.

File: examples/mnist_export_v2/export_test.py (hunk 23 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: I'm actually training for one epoch, and with big enough `batch_size` to speed up the process. I thnk for the code, that's the least amount of training I can perform. I need to dig deep to find a better way, if possible.

File: examples/mnist_export_v2/BUILD [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+py_test(
+    name = "export_test",
+    srcs = ["export_test.py"],
+    data = glob(["test_data/*.jpg"]),
+    python_version = "PY3",
+    srcs_version = "PY2AND3",
+    deps = [
+        "//tensorflow_hub:expect_numpy_installed",
+        "//tensorflow_hub:expect_tensorflow_installed",
+        ":export_lib",
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/BUILD]: This has to go the beginning of the deps list.

File: examples/mnist_export_v2/export.py [hunk 26]
```diff
@@ -0,0 +1,143 @@
+import sys
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import tensorflow_hub as hub
+import argparse
+
+"""Exporter for TF-Hub Modules in SavedModel v2.0 format.
+The module has as a single signature, accepting a batch of images with shape [None, 28, 28, 1] and returning a prediction vector.
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export.py]: Can you add a newline before "The module..." please.

File: examples/mnist_export_v2/export_test.py [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+import tempfile
+import os
+import re
+import export
+
+
+class TFHubMNISTTest(tf.test.TestCase):
+    def setUp(self):
+        self.mock_dataset = tf.data.Dataset.range(5)
+        self.mock_dataset = self.mock_dataset.map(
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export_test.py]: Please chain the dataset expressions together.

File: examples/mnist_export_v2/export_test.py [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+    def test_model_exporting(self):
+        export.train_and_export(
+            epoch=1,
+            dataset=self.mock_dataset,
+            export_path="%s/model/1" %
+            self.get_temp_dir())
+        self.assertTrue(os.listdir(self.get_temp_dir()))
+
+    def test_empty_input(self):
+        if not os.path.exists("%s/model/1" % self.get_temp_dir()):
```
Comment (vbardiovskyg) [inline examples/mnist_export_v2/export_test.py]: Remove this line

File: examples/mnist_export_v2/export_test.py (hunk 27 above)
Comment (captain-pool) [reply examples/mnist_export_v2/export_test.py]: Done

**Comment by vbardiovskyg** [comment]
Thanks for the contribution, merged in https://github.com/tensorflow/hub/commit/fb02a99bdb625271a5758a971ab23be74640674f.
